quic/transport/framesorter.nim

```diff
@@ -0,0 +1,99 @@
+import ../errors
+import std/[algorithm, strformat, tables]
+import chronos
+
+const streamEnd = high(uint64)
+
+type
+  Gap = object
+    startPos: uint64
+    endPos: uint64 # endPos is exclusive
+
+  FrameSorter* = object
+    buffer: Table[uint64, byte] # sparse byte storage
+    gaps: seq[Gap] # list of missing ranges
+    readPos: uint64 # where to emit data from
+    incoming: AsyncQueue[seq[byte]]
+    finReceived: bool # true if stream sent FIN
+
+proc initFrameSorter*(incoming: AsyncQueue[seq[byte]]): FrameSorter =
+  result.incoming = incoming
+  result.gaps = @[Gap(startPos: 0, endPos: streamEnd)] # start with [0..∞)
+  result.buffer = initTable[uint64, byte]()
+  result.readPos = 0
+
+proc normalizeGaps(gaps: var seq[Gap]) =
+  gaps.sort(
+    proc(a, b: Gap): int =
+      return cmp(a.startPos, b.startPos)
+  )
+  var merged: seq[Gap]
+  for gap in gaps:
+    if merged.len == 0 or merged[^1].endPos < gap.startPos:
+      merged.add gap
+    else:
+      merged[^1].endPos = max(merged[^1].endPos, gap.endPos)
+  gaps = merged
+
+proc removeGap(gaps: var seq[Gap], startPos, endPos: uint64) =
+  ## Remove [startPos, endPos) from gaps
+  var newGaps: seq[Gap]
+  for gap in gaps:
+    if endPos <= gap.startPos or startPos >= gap.endPos:
+      # No overlap
+      newGaps.add gap
+    else:
+      # Overlap exists
+      if startPos > gap.startPos:
+        newGaps.add Gap(startPos: gap.startPos, endPos: startPos)
+      if endPos < gap.endPos:
+        newGaps.add Gap(startPos: endPos, endPos: gap.endPos)
+  gaps = newGaps
+
+proc insert*(fs: var FrameSorter, offset: uint64, data: openArray[byte], isFin: bool) =
+  ## Insert bytes into sparse buffer
+  for i, b in data:
+    let pos = offset + uint64(i)
+    if fs.buffer.hasKey(pos):
+      if fs.buffer[pos] != b:
+        raise newException(
+          QuicError, &"conflicting byte at position {pos}. protocol violation"
+        )
+      # else: already same value, nothing to do
+    else:
+      fs.buffer[pos] = b
+
+  if isFin:
+    # Remove any data received after the end
+    fs.finReceived = true
+    removeGap(fs.gaps, offset + uint64(data.len), streamEnd)
```

Comment:
`removeGap(fs.gaps, offset + uint64(data.len), streamEnd)`
or 
`removeGap(fs.gaps, offset, streamEnd)`
??

because:

remove gap with data
`removeGap(fs.gaps, offset, offset + uint64(data.len))`
+ remove gap till stream end
`removeGap(fs.gaps, offset + uint64(data.len), streamEnd)`

Reply:
in other words, not sure about `+ uint64(data.len)` part on line 69